builtin/providers/vsphere/resource_vsphere_virtual_machine.go

```diff
@@ -0,0 +1,1062 @@
+			},
+
+			"network_interface": &schema.Schema{
+				Type:     schema.TypeList,
+				Required: true,
+				ForceNew: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"label": &schema.Schema{
+							Type:     schema.TypeString,
+							Required: true,
+							ForceNew: true,
+						},
+
+						"ip_address": &schema.Schema{
+							Type:     schema.TypeString,
+							Optional: true,
+							Computed: true,
+							ForceNew: true,
+						},
+
+						"subnet_mask": &schema.Schema{
+							Type:     schema.TypeString,
+							Optional: true,
+							Computed: true,
+							ForceNew: true,
+						},
+
+						"adapter_type": &schema.Schema{
+							Type:     schema.TypeString,
+							Optional: true,
+							ForceNew: true,
+						},
+					},
+				},
+			},
+
+			"disk": &schema.Schema{
+				Type:     schema.TypeList,
+				Required: true,
+				ForceNew: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"template": &schema.Schema{
+							Type:     schema.TypeString,
+							Optional: true,
+							ForceNew: true,
+						},
+
+						"datastore": &schema.Schema{
+							Type:     schema.TypeString,
+							Optional: true,
+							ForceNew: true,
+						},
+
+						"size": &schema.Schema{
+							Type:     schema.TypeInt,
+							Optional: true,
+							ForceNew: true,
+						},
+
+						"iops": &schema.Schema{
+							Type:     schema.TypeInt,
+							Optional: true,
+							ForceNew: true,
+						},
+					},
+				},
+			},
+
+			"boot_delay": &schema.Schema{
+				Type:     schema.TypeInt,
+				Optional: true,
+				ForceNew: true,
+			},
+		},
+	}
+}
+
+func resourceVSphereVirtualMachineCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*govmomi.Client)
+
+	vm := virtualMachine{
+		name:     d.Get("name").(string),
+		vcpu:     d.Get("vcpu").(int),
+		memoryMb: int64(d.Get("memory").(int)),
+	}
+
+	if v, ok := d.GetOk("datacenter"); ok {
+		vm.datacenter = v.(string)
+	}
+
+	if v, ok := d.GetOk("cluster"); ok {
+		vm.cluster = v.(string)
+	}
+
+	if v, ok := d.GetOk("resource_pool"); ok {
+		vm.resourcePool = v.(string)
+	}
+
+	if v, ok := d.GetOk("gateway"); ok {
+		vm.gateway = v.(string)
+	}
+
+	if v, ok := d.GetOk("domain"); ok {
+		vm.domain = v.(string)
+	}
+
+	if v, ok := d.GetOk("time_zone"); ok {
+		vm.timeZone = v.(string)
+	}
+
+	dns_suffix := d.Get("dns_suffix.#").(int)
+	if dns_suffix > 0 {
+		vm.dnsSuffixes = make([]string, 0, dns_suffix)
+		for i := 0; i < dns_suffix; i++ {
+			s := fmt.Sprintf("dns_suffix.%d", i)
+			vm.dnsSuffixes = append(vm.dnsSuffixes, d.Get(s).(string))
+		}
+	} else {
+		vm.dnsSuffixes = DefaultDNSSuffixes
+	}
+
+	dns_server := d.Get("dns_server.#").(int)
+	if dns_server > 0 {
+		vm.dnsServers = make([]string, 0, dns_server)
+		for i := 0; i < dns_server; i++ {
+			s := fmt.Sprintf("dns_server.%d", i)
+			vm.dnsServers = append(vm.dnsServers, d.Get(s).(string))
+		}
+	} else {
+		vm.dnsServers = DefaultDNSServers
+	}
```

Comment:
This is an older way of accessing list types in `ResourceData` - we now try to avoid directly interacting with the internal fields with `.`s and instead pull the whole value out using just the attribute name.

Here's an example adapted from another resource with a `TypeList` attribute:

```go
	if raw, ok := d.GetOk("dns_server"); ok {
		for i, v := range raw.([]interface{}) {
			vm.dnsServers = append(vm.dnsServers, v.(string))
		}
	}
```

For the `DefaultDNSServers` value, you can just set that in the schema itself rather than doing an else case here.

Let me know if this makes sense! :grinning: